File: src/gen.py
import os
from question import print_name

data_filename = "../docs/data.json"
template_filename = "template.html"
outdir = "../docs"

def print_dom(datum):
  return "{}->{}".format(print_name(datum["from"]), print_name(datum["to"]))

def print_doms(data):
  return ", ".join([print_dom(x) for x in data])

def print_sub_row(datum, date):
  return f"""
                  <tr>
                    <td>{print_name(datum["from"])}</td>
                    <td>{print_name(datum["to"])}</td>
                    <td>{print_info(datum, date)}</td>
                  </tr>"""

def print_sub_rows(data, date):
  return "".join([print_sub_row(x, date) for x in data])

def print_sub(datum, date):
  if "sub" in datum.keys():
    return f"""
              <table>
                <thead>
                  <tr>
                    <td>From</td>
                    <td>To</td>
                    <td>Info</td>
                  </tr>
                </thead>
                <tbody>""" + print_sub_rows(datum["sub"], date) + f"""
                </tbody>
              </table>
            """

  else:
    return "N/A"

def print_info(datum, date):
  return f'<a href="/trans/{date}/{datum["from"]}->{datum["to"]}">{datum["text"]}' if "link" in datum.keys() else datum["text"]

def print_row(datum, date):
  return f"""
          <tr>
            <td>{print_name(datum["from"])}</td>
            <td>{print_name(datum["to"])}</td>
            <td>{print_sub(datum, date)}</td>
            <td>{print_info(datum, date)}</td>
          </tr>"""

def print_row_and_date(datum, date, span=0):
  return f"""
          <tr>
            <td{' rowspan="{}"'.format(span) if span != 0 else ""}><a href="{date}">{date}</a></td>
            <td>{print_name(datum["from"])}</td>
            <td>{print_name(datum["to"])}</td>
            <td>{print_sub(datum, date)}</td>
            <td>{print_info(datum, date)}</td>
          </tr>"""

def print_rows(data, date):
  return "".join([print_row(x, date) for x in data])

def print_rows_and_date(data, date):
  return print_row_and_date(data[0], date, span=len(data)) + print_rows(data[1:], date)

def print_date_page(datum, date):
  if isinstance(datum, dict):
    return f"""<div>
      <h1>{date}: {print_dom(datum)}</h1>
      <table>
        <thead>
          <tr>
            <th>From</th>
            <th>To</th>
            <th>Sub</th>
            <th>Info</th>
          </tr>
        </thead>
        <tbody>""" + print_row(datum, date) + f"""
        </tbody>
      </table>
    </div>"""

  else:
    return f"""<div>
      <h1>{date}: {print_doms(datum)}</h1>
      <table>
        <thead>
          <tr>
            <th>From</th>
            <th>To</th>
            <th>Sub</th>
            <th>Info</th>
          </tr>
        </thead>
        <tbody>""" + print_rows(datum, date) + f"""
        </tbody>
      </table>
    </div>"""

def link(filename, link):
  with open(filename, "w") as f:
    f.write('<meta http-equiv="Refresh" content="0; {}" />'.format(link))

def links(data, date):
  d = f'{outdir}/{date}'

  if isinstance(data, dict):
    if "link" in data.keys():
      os.makedirs(f'{d}/{data["from"]}->{data["to"]}', exist_ok=True)
      link(f'{d}/{data["from"]}->{data["to"]}/index.html', "link" in data.keys())

    if "sub" in data.keys():
      for x in data["sub"]:
        if "link" in x.keys():
          os.makedirs(f'{d}/{x["from"]}->{x["to"]}', exist_ok=True)
          link(f'{d}/{x["from"]}->{x["to"]}/index.html', "link" in x.keys())

  else:
    for x in data:
      if "link" in x.keys():
        os.makedirs(f'{d}/{x["from"]}->{x["to"]}', exist_ok=True)
        link(f'{d}/{x["from"]}->{x["to"]}/index.html', "link" in x.keys())

    if "sub" in x.keys():
      for y in x["sub"]:
        if "link" in y.keys():
          os.makedirs(f'{d}/{y["from"]}->{y["to"]}', exist_ok=True)
          link(f'{d}/{y["from"]}->{y["to"]}/index.html', "link" in y.keys())

def date_pages(data, template):
  for date in data:
    os.makedirs("{}/{}".format(outdir, date), exist_ok=True)

    with open("{}/{}/index.html".format(outdir, date), "w") as f:
      f.write(template.replace("<!--REPLACE-->", print_date_page(data[date], date)))

    links(data[date], date)

def print_root_page(data):
  html = f"""<div>
      <ul class="nav">
        <li id="today"><a id="today-link" href="today">today?</a></li>
        <li><a href="crs">crs</a></li>
      </ul>
      <table>
        <thead>
          <tr>
            <th>Date</th>
            <th>From</th>
            <th>To</th>
            <th>Sub</th>
            <th>Info</th>
          </tr>
        </thead>
        <tbody>"""

  for date in data:
    if isinstance(data[date], dict):
      html += print_row_and_date(data[date], date)

    else:
      html += print_rows_and_date(data[date], date)

  html += f"""
        </tbody>
      </table>
    </div>
    <script src="trans.js"></script>"""

  return html

def main():
  import json

  with open(data_filename, "r") as f:
    data = json.load(f)

  with open(template_filename, "r") as f:
    template = f.read()

  date_pages(data, template)

  with open("{}/index.html".format(outdir), "w") as root:
    root.write(template.replace("<!--REPLACE-->", print_root_page(data)))

if __name__ == "__main__":
  main()

File: src/gen.c
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <sys/stat.h>
#include <errno.h>
#include "route.h"

#define ME "RaspBella"

#define OUTPUT_DIR "../docs"
#define INDEX "/index.html"

#define INDEX_FMT "<!DOCTYPE html>\n" \
"<html lang=\"en\">\n" \
"  <head>\n" \
"    <title>%s</title>\n" \
"    <link rel=\"stylesheet\" href=\"/main.css\">\n" \
"    <meta charset=\"UTF-8\">\n" \
"    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\n" \
"    <link rel=\"apple-touch-icon\" sizes=\"180x180\" href=\"/favicon/apple-touch-icon.png\">\n" \
"    <link rel=\"icon\" type=\"image/png\" sizes=\"32x32\" href=\"/favicon/favicon-32x32.png\">\n" \
"    <link rel=\"icon\" type=\"image/png\" sizes=\"16x16\" href=\"/favicon/favicon-16x16.png\">\n" \
"    <link rel=\"manifest\" href=\"/favicon/site.webmanifest\">\n" \
"  </head>\n" \
"  <body>\n" \
"    <ul class=\"nav\">\n" \
"      <li><a href=\"/\">home</a></li>\n" \
"      <li><a href=\"/" ME "\">about</a></li>\n" \
"      <li><a href=\"/cubing\">cubing</a></li>\n" \
"      <li><a href=\"/linux\">linux</a></li>\n" \
"      <li><a href=\"/media\">media</a></li>\n" \
"      <li><a class=\"active\" href=\"/trans\">trans</a></li>\n" \
"      <li><a href=\"/blog\">blog</a></li>\n" \
"      <li><a href=\"/calendar\">calendar</a></li>\n" \
"    </ul>\n" \
"%s\n" \
"  </body>\n" \
"%s" \
"</html>"

#define ROOT_FMT \
  "    <div>\n" \
  "      <ul class=\"nav\">\n" \
  "        <li id=\"today\"><a id=\"today_link\" href=\"today\">today?</a></li>\n" \
  "        <li><a href=\"tww\">tww</a></li>\n" \
  "        <li>\n" \
  "          <input id=\"query\" type=\"text\" placeholder=\"query...\">\n" \
  "          <button id=\"submit_query\">Submit query</button>\n" \
  "        </li>\n" \
  "      </ul>\n" \
  "      <table>\n" \
  "        <thead>\n" \
  "          <tr>\n" \
  "            <th>Date</th>\n" \
  "            <th>From</th>\n" \
  "            <th>To</th>\n" \
  "            <th>Services</th>\n" \
  "            <th>Info</th>\n" \
  "          <tr>\n" \
  "        </thead>\n" \
  "        <tbody>\n" \
  "%s\n" \
  "        </tbody>\n" \
  "      </table>\n" \
  "    </div>\n"

#define ROOT_TR_FMT \
  "          <tr>\n" \
  "            <td><a href=\"" ISO_FMT "\">" ISO_FMT "</a></td>\n" \
  "            <td>%s(%s)</td>\n" \
  "            <td>%s(%s)</td>\n" \
  "            <td>\n" \
  "              <table>\n" \
  "                <thead>\n" \
  "                  <tr>\n" \
  "                    <td>From</td>\n" \
  "                    <td>To</td>\n" \
  "                    <td>RTT</td>\n" \
  "                  </tr>\n" \
  "                </thead>\n" \
  "                <tbody>\n" \
  "%s\n" \
  "                </tbody>\n" \
  "              </table>\n" \
  "            </td>\n" \
  "            <td>%s</td>\n" \
  "          </tr>\n"

#define ROOT_SERVICES_TR_FMT \
  "                  <tr>\n" \
  "                    <td>%s(%s)</td>\n" \
  "                    <td>%s(%s)</td>\n" \
  "                    <td><a href=\"https://realtimetrains.co.uk/service/gb-nr:%s/" ISO_FMT "/detailed\">%s</a></td>\n" \
  "                  </tr>\n"

#define PAGE_FMT \
  "    <div>\n" \
  "      <h1>" ISO_FMT "</h1>\n" \
  "      <h2>%s</h2>\n" \
  "      <table>\n" \
  "        <thead>\n" \
  "          <tr>\n" \
  "            <th>From</th>\n" \
  "            <th>To</th>\n" \
  "            <th>RTT</th>\n" \
  "          <tr>\n" \
  "        </thead>\n" \
  "        <tbody>\n" \
  "%s\n" \
  "        </tbody>\n" \
  "      </table>\n" \
  "    </div>\n"

#define PAGE_TR_FMT \
  "        <tr>\n" \
  "          <td>%s(%s)</td>\n" \
  "          <td>%s(%s)</td>\n" \
  "          <td><a href=\"https://realtimetrains.co.uk/service/gb-nr:%s/" ISO_FMT "/detailed\">%s</a></td>\n" \
  "        </tr>\n"

#define ISO_FMT "%04d-%02d-%02d"
#define ISO_ARG(DATE) DATE.year, DATE.month, DATE.day

#define ARRAY_LEN(xs) sizeof(xs)/sizeof((xs)[0])

char *root(char *tbody) {
  int len = snprintf(
    NULL,
    0,
    ROOT_FMT,
    tbody
  );

  assert(!(len < 0));

  char *string = calloc(len + 1, sizeof(char));

  assert(string != NULL);

  int ret = snprintf(
    string,
    len,
    ROOT_FMT,
    tbody
  );

  assert(ret == len);

  return string;
}

char *root_services(struct route route) {
  int len = 0;
  int written = 0;

  for (int i = 0; i < route.count; ++i) {
    int from = route.services[i].from;
    int to = route.services[i].to;

    len += snprintf(
      NULL,
      0,
      ROOT_SERVICES_TR_FMT,
      crs[from].name, crs[from].code,
      crs[to].name, crs[to].code,
      route.services[i].id,
      ISO_ARG(route.on),
      route.services[i].info
    );
  }

  char *string = calloc(len + 1, sizeof(char));

  for (int i = 0; i < route.count; ++i) {
    int from = route.services[i].from;
    int to = route.services[i].to;

    written += snprintf(
      string + written,
      len - written,
      ROOT_SERVICES_TR_FMT,
      crs[from].name, crs[from].code,
      crs[to].name, crs[to].code,
      route.services[i].id,
      ISO_ARG(route.on),
      route.services[i].info
    );
  }

  return string;
}

char *root_tbody(int count, struct route *routes) {
  int len = 0;
  int written = 0;

  char **pointers = calloc(count, sizeof(char*));

  for (int i = 0; i < count; ++i) {
    pointers[i] = root_services(routes[i]);

    int from = routes[i].services[0].from;
    int to = routes[i].services[routes[i].count - 1].to;

    len += snprintf(
      NULL,
      0,
      ROOT_TR_FMT,
      ISO_ARG(routes[i].on),
      ISO_ARG(routes[i].on),
      crs[from].name, crs[from].code,
      crs[to].name, crs[to].code,
      pointers[i],
      routes[i].info
    );
  }

  char *string = calloc(len + 1, sizeof(char));

  for (int i = 0; i < count; ++i) {
    int from = routes[i].services[0].from;
    int to = routes[i].services[routes[i].count - 1].to;

    written += snprintf(
      string + written,
      len - written,
      ROOT_TR_FMT,
      ISO_ARG(routes[i].on),
      ISO_ARG(routes[i].on),
      crs[from].name, crs[from].code,
      crs[to].name, crs[to].code,
      pointers[i],
      routes[i].info
    );

    free(pointers[i]);
  }

  free(pointers);

  return string;
}

char *page(struct route route, char *tbody) {
  int len = snprintf(
    NULL,
    0,
    PAGE_FMT,
    ISO_ARG(route.on),
    route.info,
    tbody
  );

  assert(!(len < 0));

  char *string = calloc(len + 1, sizeof(char));

  assert(string != NULL);

  int ret = snprintf(
    string,
    len,
    PAGE_FMT,
    ISO_ARG(route.on),
    route.info,
    tbody
  );

  assert(ret == len);

  return string;
}

char *page_tbody(struct route route) {
  int len = 0;
  int written = 0;

  for (int i = 0; i < route.count; ++i) {
    int from = route.services[i].from;
    int to = route.services[i].to;

    len += snprintf(
      NULL,
      0,
      PAGE_TR_FMT,
      crs[from].name, crs[from].code,
      crs[to].name, crs[to].code,
      route.services[i].id,
      ISO_ARG(route.on),
      route.services[i].info
    );
  }

  char *string = calloc(len + 1, sizeof(char));

  for (int i = 0; i < route.count; ++i) {
    int from = route.services[i].from;
    int to = route.services[i].to;

    written += snprintf(
      string + written,
      len - written,
      PAGE_TR_FMT,
      crs[from].name, crs[from].code,
      crs[to].name, crs[to].code,
      route.services[i].id,
      ISO_ARG(route.on),
      route.services[i].info
    );
  }

  return string;
}

int main(void) {
  struct route routes[] = {
#include "trans.c"
  };

  FILE *fp = fopen(OUTPUT_DIR INDEX, "w");

  assert(fp != NULL);

  int count = ARRAY_LEN(routes);

  char *data = root_tbody(count, routes);

  char *body = root(data);

  free(data);

  fprintf(
    fp,
    INDEX_FMT,
    "transportation - " ME,
    body,
    "  <script src=\"trans.js\"></script>\n"
  );

  free(body);

  char filename[] = OUTPUT_DIR "/yyyy-mm-dd" INDEX;

  for (int i = 0; i < count; ++i) {
    int ret = snprintf(
      filename + sizeof(OUTPUT_DIR),
      11,
      ISO_FMT,
      ISO_ARG(routes[i].on)
    );

    assert(ret == 10);

    if (mkdir(filename, 0755) != 0) {
      assert(errno == EEXIST);
    }

    filename[sizeof(OUTPUT_DIR) + 10] = '/';

    fp = freopen(filename, "w", fp);

    filename[sizeof(OUTPUT_DIR) + 10] = 0;

    char title[] = "yyyy-mm-dd - transportation - " ME;

    ret = snprintf(
      title,
      sizeof(title),
      ISO_FMT " - transportation - " ME,
      ISO_ARG(routes[i].on)
    );

    data = page_tbody(routes[i]);

    body = page(routes[i], data);

    free(data);

    fprintf(
      fp,
      INDEX_FMT,
      title,
      body,
      ""
    );

    free(body);
  }

  fp = freopen(OUTPUT_DIR "/data.json", "w", fp);

  assert(fp != NULL);

  fprintf(fp, "{");

  for (int i = 0; i < count; ++i) {
    fprintf(
      fp,
      "\"" ISO_FMT "\":{"
      "\"info\":\"%s\","
      "\"services\":[",
      ISO_ARG(routes[i].on),
      routes[i].info
    );

    for (int j = 0; j < routes[i].count; ++j) {
      fprintf(
        fp,
        "{\"info\":\"%s\","
        "\"id\":\"%s\","
        "\"from\":\"%s\","
        "\"to\":\"%s\"}",
        routes[i].services[j].info,
        routes[i].services[j].id,
        crs[routes[i].services[j].from].code,
        crs[routes[i].services[j].to].code
      );

      if (j < (routes[i].count - 1)) {
        fprintf(fp, ",");
      }
    }

    fprintf(
      fp,
      "]}"
    );

    if (i < (count - 1)) {
      fprintf(fp, ",");
    }
  }

  fprintf(fp, "}");

  fclose(fp);
}
